api/client/batch_client.py

```diff
@@ -40,51 +45,67 @@ def get_data(self, url, headers, params=None):
         while retry_count < cfg.MAX_RETRIES:
             start_time = time.time()
             http_request = HTTPRequest(url, method="GET", headers=headers,
-                                       request_timeout=cfg.TIMEOUT, connect_timeout=cfg.TIMEOUT)
-            data = yield self._http_client.fetch(http_request)
-            elapsed_time = time.time() - start_time
-            log_record = dict(base_log_record)
-            log_record['elapsed_time_in_ms'] = 1000 * elapsed_time
-            log_record['retry_count'] = retry_count
-            log_record['status_code'] = data.code
-            if data.code == 200:
+                                       request_timeout=cfg.TIMEOUT,
+                                       connect_timeout=cfg.TIMEOUT)
+            try:
+                data = yield self._http_client.fetch(http_request)
+                elapsed_time = time.time() - start_time
+                log_record = dict(base_log_record)
+                log_record['elapsed_time_in_ms'] = 1000 * elapsed_time
+                log_record['retry_count'] = retry_count
+                log_record['status_code'] = data.code
                 self._logger.debug('OK', extra=log_record)
                 raise gen.Return(data.body)
-            retry_count += 1
-            if retry_count < cfg.MAX_RETRIES:
-                self._logger.warning(data.text, extra=log_record)
-            else:
-                self._logger.error(data.text, extra=log_record)
-                raise Exception('Giving up on {} after {} tries. Error is: {}.'.format(
-                    url, retry_count, data.text))
+            except HTTPError as e:
+                elapsed_time = time.time() - start_time
+                log_record = dict(base_log_record)
+                log_record['elapsed_time_in_ms'] = 1000 * elapsed_time
+                log_record['retry_count'] = retry_count
+                log_record['status_code'] = e.code
+                if retry_count < cfg.MAX_RETRIES:
+                    self._logger.warning(e.response.error, extra=log_record)
+                    retry_count += 1
+                    if e.code == 429:
+                        rate_limit_lock.clear()
```

Comment:
the rate_limit_lock event is not being yielded (await in py3), so it's not doing anything. 

tornado is a single-thread async library so a vanilla python time.sleep() will actually block all execution. this is a fine assumption if we are just doing a batch async set of calls (we want a 429 event to stop all our requests that are in flight). **i propose remove the lock for now and just keep the sleep part here**

however if someone wants to combine our async calls with other async coroutines then we should indeed 
```_ = yield rate_limit_lock```
here I believe, and make sure it's initially set, then follow it with a [tornado.gen.sleep](https://www.tornadoweb.org/en/stable/gen.html#tornado.gen.sleep)
**but let's do this down the line**

an additional issue is that if we have multiple requests in flight receiving a 429 we want all of them to wait, instead of one by one processing this backoff, i.e. maybe having a shared variable across the co-routines. but this is even less of a priority.

Reply:
Ah I didn't realize sleep would block all coroutines. Will remove.

Regarding your point about the multiple requests in flight, I think this is what you were getting at, but to try to clarify: several coroutines could potentially run into a rate limit within the same period before the first response is received, so ideally we could consider request start time and consolidate 429s within the same period and just do one backoff instead of several in a row when processing a backlog of responses.

Reply:
correct. yep let's leave that aside as premature optimization for now.